buffer/src/main/java/io/netty/buffer/ByteBufUtil.java

```diff
@@ -245,9 +245,9 @@ public static int indexOf(ByteBuf needle, ByteBuf haystack) {
             return 0;
         }
 
-        // When the needle has only one byte that can be read,
-        // the firstIndexOf method needs to be called
-        if (m == 1) {
+        // When the needle has only one byte that can be read and haystack is AbstractByteBuf,
+        // the firstIndexOf method can be used
+        if (m == 1 && haystack instanceof AbstractByteBuf) {
             return firstIndexOf((AbstractByteBuf) haystack, haystack.readerIndex(),
                     haystack.writerIndex(), needle.getByte(needle.readerIndex()));
```

Comment:
You can use the `ByteBuf.indexOf(int fromIndex, int toIndex, byte value)` method instead. When the buffer is an `AbstractByteBuf` this will delegate to `firstIndexOf`.

Then the shortcut won't be skipped for other buffer implementations.

Reply:
Thanks